source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,565 @@
+    // value of the operand might not be fully resolved yet, so this
+    // OpSpecConstantOp instruction can not be folded.
+    return nullptr;
+  }
+}
+
+std::unique_ptr<Value>
+FoldSpecConstantOpAndCompositePass::GetVectorShuffleResult(
+    ir::Instruction* inst) {
+  assert(inst->NumInOperands() - 1 > 2 &&
+         "OpSpecConstantOp instruction with VectorShuffle spec opcode requires "
+         "more than 2 operands (2 vector ids and at least one literal operand");
+
+  std::vector<uint32_t> concatenated_id_vec;
+  for (uint32_t i : std::vector<uint32_t>{1, 2}) {
+    assert(inst->GetInOperand(i).type == SPV_OPERAND_TYPE_ID &&
+           "The vector operand must have a SPV_OPERAND_TYPE_ID type");
+    auto iter = id_to_const_val_.find(inst->GetSingleWordInOperand(i));
+    if (iter != id_to_const_val_.end()) {
+      assert(iter->second->AsComposite() &&
+             "The vector operand must have a composite type constant value");
+      CompositeValue* vec_val = iter->second->AsComposite();
+      for (uint32_t nested_const_id : vec_val->nested) {
+        assert(id_to_const_val_.find(nested_const_id) !=
+                   id_to_const_val_.end() &&
+               "Front-end composite constants should only contain front-end "
+               "constants");
+        assert(
+            id_to_const_val_.find(nested_const_id)->second->AsScalar() &&
+            "The vector operand must have a vector type constant value, which "
+            "means its nested values must be scalar type constants");
+        concatenated_id_vec.push_back(nested_const_id);
+      }
+    } else {
+      // Does not find the operand in the id_to_const_val_ map, the
+      // value of the operand might not be fully resolved yet, so this
+      // OpSpecConstantOp instruction can not be folded.
+      return nullptr;
+    }
+  }
+  auto result = std::unique_ptr<Value>(new CompositeValue);
+  for (uint32_t i = 3; i < inst->NumInOperands(); i++) {
+    assert(inst->GetInOperand(i).type == SPV_OPERAND_TYPE_LITERAL_INTEGER &&
+           "The literal operand must have type "
+           "SPV_OPERAND_TYPE_LITERAL_INTEGER");
+    uint32_t literal = inst->GetSingleWordInOperand(i);
+    assert(literal < concatenated_id_vec.size() &&
+           "Literal index out of bound of the concatenated vector");
+
+    result->AsComposite()->nested.push_back(concatenated_id_vec[literal]);
+  }
+  return result;
+}
+
+std::tuple<std::unique_ptr<Value>, std::vector<uint32_t>>
+FoldSpecConstantOpAndCompositePass::GetBroadcastOperationResult(
+    ir::Instruction* inst) {
+  const uint32_t spec_opcode_count = 1;
+  uint32_t num_operands = inst->NumInOperands();
+  uint32_t operand_index = 1;
+  uint32_t operand_id = inst->GetSingleWordInOperand(operand_index);
+  auto iter = id_to_const_val_.find(operand_id);
+  if (iter == id_to_const_val_.end()) {
+    return std::make_tuple(nullptr, std::vector<uint32_t>());
```

Comment:
That will need initializer_list<uint32_t> so I think this won't save any words....